File: hostd/src/api/routes/ports.rs
//! Exposed-port routes: the per-VM registry of guest ports (with labels) that
//! HTTP workloads listen on, plus minting of the ephemeral JWTs the proxy
//! server uses to authenticate forwarded requests.
//!
//! The registry lives in `VmConfig.network_config.exposed_ports`; the proxy
//! (`hostd/src/proxy/`) re-validates it on every forwarded request, so
//! removing a port here revokes proxy access immediately.

use axum::{
    Json, Router,
    extract::{Path, State},
    http::StatusCode,
    routing::{get, post},
};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::{
    api::{
        error::{ApiJson, ApiResult},
        server::AppState,
    },
    error::Error,
    net::ExposedPort,
    proxy::{DEFAULT_TTL_SECS, Proto},
    vmm::vm::VmId,
};

/// Exposed-port routes, to be nested under `/vms/{id}/ports`.
pub(crate) fn routes() -> Router<AppState> {
    Router::new()
        .route("/", get(list_exposed_ports).post(add_exposed_port))
        .route("/{port}", axum::routing::delete(remove_exposed_port))
        .route("/{port}/token", post(mint_port_token))
}

/// List a vm's exposed ports.
pub(crate) async fn list_exposed_ports(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> ApiResult<Json<Vec<ExposedPort>>> {
    Ok(Json(state.vmm.list_exposed_ports(&VmId(id)).await?))
}

/// Expose a guest port, with a label describing its purpose.
pub(crate) async fn add_exposed_port(
    State(state): State<AppState>,
    Path(id): Path<String>,
    ApiJson(port): ApiJson<ExposedPort>,
) -> ApiResult<(StatusCode, Json<ExposedPort>)> {
    let port = state.vmm.add_exposed_port(&VmId(id), port).await?;
    Ok((StatusCode::CREATED, Json(port)))
}

/// Remove an exposed port by port number.
pub(crate) async fn remove_exposed_port(
    State(state): State<AppState>,
    Path((id, port)): Path<(String, u16)>,
) -> ApiResult<StatusCode> {
    state.vmm.remove_exposed_port(&VmId(id), port).await?;
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Debug, Deserialize)]
pub(crate) struct MintTokenRequest {
    /// Requested token lifetime in seconds; defaults to
    /// [`DEFAULT_TTL_SECS`] and is clamped to the proxy's maximum.
    #[serde(default)]
    ttl_secs: Option<u64>,
    /// Forwarding mode the token is valid for: `http` (default) or `tcp`
    /// (e.g. the Postgres wire protocol, see `proxy/tcp.rs`).
    #[serde(default)]
    proto: Option<Proto>,
}

#[derive(Debug, Serialize)]
pub(crate) struct MintTokenResponse {
    token: String,
    expires_at: DateTime<Utc>,
}

/// Mint an ephemeral JWT authorizing proxy requests to this exposed port.
pub(crate) async fn mint_port_token(
    State(state): State<AppState>,
    Path((id, port)): Path<(String, u16)>,
    ApiJson(body): ApiJson<MintTokenRequest>,
) -> ApiResult<(StatusCode, Json<MintTokenResponse>)> {
    // Minting requires the port to be exposed right now; the proxy
    // re-checks this on every forwarded request as well.
    let vm_id = VmId(id);
    let ports = state.vmm.list_exposed_ports(&vm_id).await?;
    if !ports.iter().any(|p| p.port == port) {
        return Err(Error::PortNotExposed {
            vm_id: vm_id.to_string(),
            port,
        }
        .into());
    }
    let (token, expires_at) = state.tokens.mint(
        &vm_id,
        port,
        body.proto.unwrap_or(Proto::Http),
        body.ttl_secs.unwrap_or(DEFAULT_TTL_SECS),
    )?;
    Ok((
        StatusCode::CREATED,
        Json(MintTokenResponse { token, expires_at }),
    ))
}

File: clients/node/src/ports.ts
import { HttpClient } from './http.js';
import type { ExposedPort } from './types.js';

/**
 * Forwarding mode a proxy token is valid for (hostd's `proxy::Proto`).
 * `http` reverse-proxies requests; `tcp` splices raw bytes instead (e.g. the
 * Postgres wire protocol, where the JWT rides in the `tikovm_token` startup
 * parameter).
 */
export type ProxyProto = 'http' | 'tcp';

export interface MintPortTokenOptions {
  /** Requested token lifetime in seconds; hostd clamps it to its maximum. */
  ttl_secs?: number;
  /** Forwarding mode the token is valid for; defaults to `http`. */
  proto?: ProxyProto;
}

/** Response of POST `/api/vms/{id}/ports/{port}/token`. */
export interface PortToken {
  token: string;
  expires_at: string;
}

/**
 * The `/api/vms/{id}/ports` resource: the per-VM registry of exposed guest
 * ports (with labels) plus minting of the ephemeral JWTs the proxy server
 * uses to authenticate forwarded requests. Unexposing a port revokes proxy
 * access immediately (hostd re-validates it on every connection).
 */
export class PortsApi {
  /** @internal */
  constructor(
    private readonly http: HttpClient,
    private readonly vmId: string,
  ) {}

  /** List the VM's exposed ports. */
  async list(): Promise<ExposedPort[]> {
    return this.http.request<ExposedPort[]>({
      method: 'GET',
      path: `/api/vms/${this.vmId}/ports`,
    });
  }

  /** Expose a guest port, with a label describing its purpose. */
  async expose(port: ExposedPort): Promise<ExposedPort> {
    return this.http.request<ExposedPort>({
      method: 'POST',
      path: `/api/vms/${this.vmId}/ports`,
      body: port,
    });
  }

  /** Stop exposing a port by number; proxy access to it is revoked immediately. */
  async remove(port: number): Promise<void> {
    await this.http.request<void>({
      method: 'DELETE',
      path: `/api/vms/${this.vmId}/ports/${port}`,
    });
  }

  /**
   * Mint an ephemeral JWT authorizing proxy requests to this exposed port.
   * Requires the port to be currently exposed.
   */
  async token(port: number, options: MintPortTokenOptions = {}): Promise<PortToken> {
    const body: MintPortTokenOptions = {};
    if (options.ttl_secs !== undefined) {
      body.ttl_secs = options.ttl_secs;
    }
    if (options.proto !== undefined) {
      body.proto = options.proto;
    }
    return this.http.request<PortToken>({
      method: 'POST',
      path: `/api/vms/${this.vmId}/ports/${port}/token`,
      body,
    });
  }
}
